src/dma/traits.rs

```diff
@@ -0,0 +1,1394 @@
+use super::*;
+use crate::{
+    bb,
+    pac::{self, DMA1, DMA2, RCC},
+};
+use core::ops::Deref;
+
+pub(crate) mod sealed {
+    /// Converts value to bits for setting a register value.
+    pub trait Bits<T> {
+        /// Returns the bit value.
+        fn bits(self) -> T;
+    }
+    pub trait Sealed {}
+}
+use sealed::{Bits, Sealed};
+
+/// Trait for DMA streams types.
+pub trait Stream: Sealed {
+    /// Number of the register stream.
+    const NUMBER: usize;
+
+    /// Clear all interrupts for the DMA stream.
+    fn clear_interrupts(&mut self);
+
+    /// Clear transfer complete interrupt (tcif) for the DMA stream.
+    fn clear_transfer_complete_interrupt(&mut self);
+
+    /// Clear half transfer interrupt (htif) for the DMA stream.
+    fn clear_half_transfer_interrupt(&mut self);
+
+    /// Clear transfer error interrupt (teif) for the DMA stream.
+    fn clear_transfer_error_interrupt(&mut self);
+
+    /// Clear direct mode error interrupt (dmeif) for the DMA stream.
+    fn clear_direct_mode_error_interrupt(&mut self);
+
+    /// Clear fifo error interrupt (feif) for the DMA stream.
+    fn clear_fifo_error_interrupt(&mut self);
+
+    /// Get transfer complete flag.
+    fn get_transfer_complete_flag() -> bool;
+
+    /// Get half transfer flag.
+    fn get_half_transfer_flag() -> bool;
+
+    /// Set the peripheral address (par) for the DMA stream.
+    fn set_peripheral_address(&mut self, value: u32);
+
+    /// Set the memory address (m0ar) for the DMA stream.
+    fn set_memory_address(&mut self, value: u32);
+
+    /// Set the double buffer address (m1ar) for the DMA stream.
+    fn set_memory_double_buffer_address(&mut self, value: u32);
+
+    /// Set the number of transfers (ndt) for the DMA stream.
+    fn set_number_of_transfers(&mut self, value: u16);
+
+    /// Enable the DMA stream.
+    ///
+    /// # Safety
+    ///
+    /// The user must ensure that all registers are properly configured.
+    unsafe fn enable(&mut self);
+
+    /// Returns the state of the DMA stream.
+    fn is_enabled() -> bool;
+
+    /// Disable the DMA stream.
+    fn disable(&mut self);
+
+    /// Set the channel for the (chsel) the DMA stream.
+    fn set_channel<C: Channel>(&mut self, channel: C);
+
+    /// Set the priority (pl) the DMA stream.
+    fn set_priority(&mut self, priority: config::Priority);
+
+    /// Set the memory size (msize) for the DMA stream.
+    ///
+    /// # Safety
+    /// This must have the same alignment of the buffer used in the transfer.
+    /// Valid values:
+    ///     * 0 -> byte
+    ///     * 1 -> half word
+    ///     * 2 -> word
```

Comment:
Those are the bits used to write to the register, and the only valid values.

Reply:
I see, this API is more low-level than I initially figured.
You could make this clearer by taking a dedicated enum that implemented `Bits` to get these values. Also would avoid the need to check if the value is valid.

Reply:
There is no check currently, the enum based would need a check when creating the enum though, since we got the size from `mem::align_of`. Not sure if I want to do that since it doesn't seem necessary and the API is already unsafe.

Reply:
It wouldn't be necessary, just more readable. But if you want to avoid that additional complexity that's okay with me.